### PR TITLE
[WIP] Add support for Carthage through xcodegen

### PR DESCRIPTION
**Pull request by JonasVautherin**
Trying to implement a solution for #329.

FYI: @MrMage, @sbarow, @ishkawa

I get an error saying:

```
Showing All Messages
:-1: Multiple commands produce '/Users/jones/Library/Developer/Xcode/DerivedData/SwiftGRPC-XcodeGen-gohnuovawbgognemswwzuzdzqfeh/Build/Products/Debug-iphonesimulator/BoringSSL.framework/Headers/internal.h':
1) Target 'BoringSSL_iOS' (project 'SwiftGRPC-XcodeGen') has copy command from '/tmp/test_xcodegen/grpc-swift/Sources/BoringSSL/crypto/bio/internal.h' to '/Users/jones/Library/Developer/Xcode/DerivedData/SwiftGRPC-XcodeGen-gohnuovawbgognemswwzuzdzqfeh/Build/Products/Debug-iphonesimulator/BoringSSL.framework/Headers/internal.h'
2) Target 'BoringSSL_iOS' (project 'SwiftGRPC-XcodeGen') has copy command from '/tmp/test_xcodegen/grpc-swift/Sources/BoringSSL/crypto/bytestring/internal.h' to '/Users/jones/Library/Developer/Xcode/DerivedData/SwiftGRPC-XcodeGen-gohnuovawbgognemswwzuzdzqfeh/Build/Products/Debug-iphonesimulator/BoringSSL.framework/Headers/internal.h'
...
[and a lot more similar ones]
```

I'm stuck there, I have absolutely no idea what's happening.

### CODE REVIEW

**Comment by MrMage** [comment]
Thank you! Could you please also add tooling to generate the Xcode project? Also, I assume the Carthage project will also need to be committed?

In terms of the error, my suspicion would be that those internal headers should not be marked "public" to avoid duplicating them. I think this is one of the reasons why the [BoringSSL Podspec](https://github.com/grpc/grpc/blob/master/src/objective-c/BoringSSL-GRPC.podspec) is fairly elaborate.

File: project.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      - target: CgRPC_$platform
+      - carthage: SwiftProtobuf
+    type: framework
+  BoringSSL:
+    platform: [iOS]
+    settings:
+      SWIFT_VERSION: 4
+    scheme: {}
+    sources:
+        - Sources/BoringSSL
```
Comment (sbarow) [inline project.yml]: If you change this to 
```
    sources:
      - path: Sources/BoringSSL
        headerVisibility: project 
        excludes:
          - "README.md"
```

It should solve the issue with the headers.

**Comment by JonasVautherin** [comment]
Thanks a lot, now BoringSSL builds! Moving to CgRPC, I get the following errors: 

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:70:9: error: no member named 'memcpy' in the global namespace
using ::memcpy;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:71:9: error: no member named 'memmove' in the global namespace
using ::memmove;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:72:9: error: no member named 'strcpy' in the global namespace
using ::strcpy;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:73:9: error: no member named 'strncpy' in the global namespace
using ::strncpy;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:74:9: error: no member named 'strcat' in the global namespace
using ::strcat;
      ~~^
...
[and a lot of similar ones]
```

Any idea?

**Comment by JonasVautherin** [comment]
Some progress here for CgRPC:
* All the headers should apparently not go in the "project" section of the "Headers build phase"
* cgrpc.h should go in the "public" section of the "Headers build phase"
* In "Compile Sources" (still "Build Phases"), the files appear twice: once from Source/cgrpc/include/src, and once from Sources/cgrpc/src, resulting in "duplicated symbol" errors. I guess the includes should be excluded from "Compile Sources"
* libz should apparently be linked

@sbarow: I am trying to do that in the project.yml, but I'm not very successful so far. So I'm open to suggestions :-)

**Comment by JonasVautherin** [comment]
Well, I'm lost. I don't know why I can't make CgRPC build, I don't know if xcodegen can do what I need at all, I spent way too much time on that and I'm fed up.

**Pull request closed without merging**